### DISCUSSION TITLE
SML VBUS RemehaSol B/2 - negative Temperature values

### DISCUSSION THREAD

**Discussion by FreddyFFM** [arendst/Tasmota#13628] [Q&A] [answered]
Hi,

i already searched through a lot of sites and read the SML docu several times, but I did not find a solution on how to proceed with negative Temperature values. I took the example script and took the values my RemehaSol B/2 provides via VBUS:

>D
r="1,aa10005d101000010a67"
>B
=>sensor53 r
>M 1
+1,3,v,0,9600,Solar
%r%vo0uw@10,Kollektor,°C,kol,10
%r%vo2uw@10,Speicher unten,°C,spu,1
%r%vo4uw@10,Speicher oben,°C,spo,1
%r%vo8ub@1,Pumpe,%%,pump,0
#

But when the Collector temperature turns into negative (during the night), the output is a temperature above 6000. Referring to e.g. https://github.com/bbqkees/vbus-arduino-domoticz/blob/master/Arduino-Code/ArduinoVBusDecoder.ino (line 840), there needs some calculation to be done, but I did not figure out how to do it within the script.
Anybody who can help here?

**Comment by gemu2015** [comment, marked as answer]
you specify uw for your collector value. this means unsigned word.
choose sw which means signed word.

**Reply by FreddyFFM** [reply]
I tried sw, but this gives me total different values (positive and negative).
Using sb gives me the correct negative values but the wrong positives ones.

The above mentioned script uses this calculation:

>float CalcTemp(int Byte1, int Byte2) {
   int v;
    v = Byte1 << 8 | Byte2; //bit shift 8 to left, bitwise OR
    if (Byte1 == 0x00){
    v= v & 0xFF;  
    }
    if (Byte1 == 0xFF)
        v = v - 0x10000;
    if (v==SENSORNOTCONNECTED)
        v=0;
    return (float)((float) v * 0.1);
    }
    
   Is there a way to script this calculation?

**Reply by gemu2015** [reply]
strange coding but you can mimic that:

add to user config
#define USE_SML_SCRIPT_CMD

then add to script
e.g
\>D
coltemp=0
Byte1=0
Byte2=0

\>S
; decompose bytes
coltemp=sml[1]
Byte1=(coltemp>>8)&0xff
Byte2=coltemp&0xff
; now calculate as in your source example
if Byte1==0x00 {
; this is actually useless, since high byte is zero already
coltemp=coltemp&0xff
}
if Byte1==0xff {
; this is the strange coding for negative values
coltemp=coltemp-0x10000
}

; show in console
print collector temp = %coltemp%


if that works you can hide wrong value in webui (see sml docs)

and add new value to webui and mqtt (see scripting docs)

**Reply by FreddyFFM** [reply]
Thanks for sharing this approach, I embeded the script and calculation works (so far I could not finally test as it was above 0° the past days).
But I am still struggeling with finding the right way to publish the calculated value to MQTT. I went through the docs and was able to hide the original value but not yet found the correct hint to publish.

**Reply by gemu2015** [reply]
add #define USE_SCRIPT_JSON_EXPORT
and add >J section

example:

; transmit these values with MQTT
\>J
"daily_consumption":%pi_d%,"daily_feed":%po_d%

**Reply by FreddyFFM** [reply]
Thanks for your support, values are published now, need to wait for cold temperatures now 👍
